pkg/reconciler/events/auditlogs/auditlogs.go

```diff
@@ -70,25 +72,10 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, s *v1alpha1.CloudAuditLo
 	s.Status.InitializeConditions()
 	s.Status.ObservedGeneration = s.Generation
 
-	// If GCP ServiceAccount is provided, configure workload identity.
-	if s.Spec.ServiceAccount != nil {
-		// Create corresponding k8s ServiceAccount if doesn't exist, and add ownerReference to it.
-		kServiceAccountName := psresources.GenerateServiceAccountName(s.Spec.ServiceAccount)
-		kServiceAccount, err := c.serviceAccountLister.ServiceAccounts(s.Namespace).Get(kServiceAccountName)
-		if err != nil {
-			if !apierrs.IsNotFound(err) {
-				logging.FromContext(ctx).Desugar().Error("Failed to get k8s service account", zap.Error(err))
-				return reconciler.NewEvent(corev1.EventTypeWarning, workloadIdentityFailedReason, "Getting k8s service account failed with: %s", err)
-			}
-			kServiceAccount = nil
-		}
-		kServiceAccount, event := c.CreateServiceAccount(ctx, s, kServiceAccount)
-		if event != nil {
-			return event
-		}
-		// Add iam policy binding to GCP ServiceAccount.
-		if err := psresources.AddIamPolicyBinding(ctx, s.Spec.Project, s.Spec.ServiceAccount, kServiceAccount); err != nil {
-			return reconciler.NewEvent(corev1.EventTypeWarning, workloadIdentityFailedReason, "Adding iam policy binding failed with: %s", err)
+	// If GCP ServiceAccount is provided, reconcile workload identity.
+	if s.Spec.ServiceAccount != "" {
+		if _, err := c.Identity.ReconcileWorkloadIdentity(ctx, s.Spec.Project, s.Namespace, s); err != nil {
```

Comment:
I think you don't need to pass s.Namespace... you can get it from OwnerReferable..

`s.GetObjectMeta().GetNamespace()`

Reply:
are you sure s.Spec.Project is not empty? I don't recall much, but I thought we were setting the Status.ProjectID with the value... It's better if its s.Spec.Project, but just double check

Reply:
The projectID is used in `setIamPolicy()`. It can be empty, and handled in `setIamPolicy()` with `projectId, err := utils.ProjectID(projectID)`, it will get `metadata.ProjectID()` when input projectID is empty